### PR TITLE
Automation of the registration of clusters into Argo CD  following a GitOps approach

### PR DESCRIPTION
**Pull request by fgiloux**
Goal: Providing a GitOps approach to automate the registration of clusters into Argo CD.

Content: 
- a PipelineRun integrated into Pipelines as Code, which automate the registration from a repo push or PR
- a container image for supporting the Task
- a script with the necessary logic
- manifests for namespaces and rbac
- documentation

Signed-off-by: Frederic Giloux <fgiloux@redhat.com>

### CODE REVIEW

**Comment by fgiloux** [comment]
@Roming22 a couple of points I would like to discuss before this PR gets merged:
- `credentials/kubeconfig/plnsvc`: why plnsvc as name?  How does it identify the role the components under the directory are playing within the service? I would have expected more something like "workload" (very overloaded), "compute" or "target".
- `/environment/plnsvc/cluster-name/config: do we need `config`? I would expect everything there to be configuration. What have you got in mind?

File: gitops/environment/plnsvc/base/ns-rbac/tekton-ns.yaml
Comment (bnallapeta) [inline gitops/environment/plnsvc/base/ns-rbac/tekton-ns.yaml:1]: Typo: **Tekton**

File: gitops/argocd/image/register.sh
Comment (bnallapeta) [inline gitops/argocd/image/register.sh:128]: If our intention is to just create the namespaces in the first run, we could create another directory for handling this better.

- pipelines-service/gitops/environment/plnsvc/base/
  - ns
    - kcp-syncer-ns.yaml
    - management-ns.yaml
    - tekton-ns.yaml
    - kustomization.yaml
  - rbac
    - argo-syncer-role.yaml
    - argo-tekton-role.yaml
    - argo-syncer-rolebinding.yaml
    - argo-tekton-rolebinding.yaml
    - kustomization.yaml

With this kind of a directory structure, lines 129 and 131 could call kustomization.yaml under ns and rbac respectively.

File: gitops/argocd/image/register.sh
Comment (bnallapeta) [inline gitops/argocd/image/register.sh:87]: Add double quote around array expansion.
"${files[@]}"

File: gitops/argocd/image/register.sh
Comment (bnallapeta) [inline gitops/argocd/image/register.sh:58]: missing opening braces
if [[ -z "${DATA_DIR}" ]]; then

File: gitops/argocd/image/register.sh
Comment (bnallapeta) [inline gitops/argocd/image/register.sh:52]: missing opening braces
if [[ -z "${ARGO_PWD}" ]]; then

File: gitops/argocd/image/Dockerfile
Comment (bnallapeta) [inline gitops/argocd/image/Dockerfile:19]: If we are not passing any commands, why the empty CMD [""]
(since we already have ENTRYPOINT in the above instruction, we could remove the CMD instruction)

File: gitops/argocd/README.md
Comment (bnallapeta) [inline gitops/argocd/README.md:17]: If we need to create another folder called gitops/credentials to store the kubeconfigs, then its better that we create the folder gitops/credentials/kubeconfig/ and have a .gitignore file or a README in here. Users can therefore commit just the kubeconfig files in their future PRs.

File: gitops/argocd/README.md
Comment (bnallapeta) [inline gitops/argocd/README.md:7]: Typo: Pipelines **as** Code

**Review by bnallapeta — COMMENTED** [review]
Did not run it locally yet. Few cosmetic and structural changes.

File: gitops/argocd/README.md
Comment (bnallapeta) [inline gitops/argocd/README.md:20]: bases is deprecated. Need to use **resources**.

File: gitops/argocd/image/register.sh
Comment (fgiloux) [reply gitops/argocd/image/register.sh:128]: Yes I thought of that. I was not sure of what would be better as it breaks a logical unit into peaces due to the way the Argo CD client handles things. I can do the split if you like ti better

Comment (bnallapeta) [reply gitops/argocd/image/register.sh:128]: Hmm agreed. But since we are going the kustomize way, let's do it that way. It creates a neater way to add/remove stuff in the future too.
(I am doing something similar for "ArgoCD in kubernetes vs openshift" too.)

File: gitops/argocd/README.md
Comment (fgiloux) [reply gitops/argocd/README.md:17]: yes, actually I had created one, forgot to push it

**Comment by fgiloux** [comment]
@bnallapeta Thanks for the review. A lot of good points. I have amended the PR and I believe they are all addressed. Let me know how you see it.

File: gitops/argocd/README.md
Comment (bnallapeta) [inline gitops/argocd/README.md:17]: 2 kustomization files now under two folders, viz., namespaces and argocd-rbac.

- argocd-rbac
  - kustomization.yaml
- namespaces.yaml
  - kustomization.yaml

File: gitops/argocd/image/register.sh
Comment (bnallapeta) [inline gitops/argocd/image/register.sh:130]: Currently, clustername is passed for 'argocd add' which is incorrect. We need to add context name. Modified this line to extract context name. But I think you can do this in the above for loop too and store an array of context names too.

```
KUBECONFIG=${DATA_DIR}/gitops/credentials/kubeconfig/plnsvc/${kubeconfigs[$i]} argocd ${insecure} cluster add "$(yq ".contexts[$i].name" <"${DATA_DIR}/gitops/credentials/kubeconfig/plnsvc/${kubeconfigs[$i]}")" --system-namespace argocd-management --namespace=tekton-pipelines --namespace=kcp-syncer --yes
```

File: gitops/argocd/README.md
Comment (bnallapeta) [inline gitops/argocd/README.md:17]: we need to also think about the user experience here. Since the name of the cluster and the name of this folder has to be same, user has to end up creating a folder name like this. Is there any other way to overcome this?

`gitops/environment/plnsvc/api-ci-ln-d5t7nx2-72292-origin-ci-int-gce-dev-rhcloud-com/`

File: gitops/argocd/image/register.sh
Comment (fgiloux) [reply gitops/argocd/image/register.sh:88]: actually not the .gitignore as it is a hidden file but I will move both one level up

**Comment by Roming22** [comment]
> @Roming22 a couple of points I would like to discuss before this PR gets merged:
> 
> * `credentials/kubeconfig/plnsvc`: why plnsvc as name?  How does it identify the role the components under the directory are playing within the service? I would have expected more something like "workload" (very overloaded), "compute" or "target".

I'm trying to split the configuration file per resource type. I used `plnsvc` to make it obvious that the kubeconfig referenced it the directory would host the Pipeline Service service. A more generic name would not necessarily convey that information.

> * `/environment/plnsvc/cluster-name/config: do we need `config`? I would expect everything there to be configuration. What have you got in mind?
Every cluster has 2 directories (`config` and `gitops`). For `plnsvc` clusters, `config` holds the information that ties the cluster to argocd, in the case of a `kcp` cluster there's additional information about tenants and namespaces. Those files holds the configuration that are the input to the pipeline_as_code process.
`gitops` holds the argocd manifests.

File: gitops/argocd/README.md
Comment (fgiloux) [reply gitops/argocd/README.md]: I am open to proposals but:
- relying on ordering is a no go
- using a key may be an option but I am not sure it improves the UX

File: gitops/argocd/image/register.sh
Comment (fgiloux) [reply gitops/argocd/image/register.sh:88]: I still needed a second .gitignore so that the directory is not empty and I am able to commit it

File: gitops/argocd/README.md
Comment (Roming22) [inline gitops/argocd/README.md:39]: How about switching to storing all credentials in `credentials/kubeconfig`, but encrypted?

File: gitops/argocd/image/register.sh
Comment (Roming22) [inline gitops/argocd/image/register.sh:30]: `cluster files` -> `environment files`?

File: gitops/argocd/image/register.sh
Comment (Roming22) [inline gitops/argocd/image/register.sh:78]: plnsvc is the cluster type, like KCP. So I would expect as many kubeconfig under that dir as plnsvc clusters. E.g.
```
plnsvc/
- my-aws-cluster0.yaml
- my-azure-cluster0.yaml
- my-azure-cluster1.yaml
- my-clusterbot-cluster0.yaml
```
I don't think we need additional subdirs, but I could be wrong.

File: gitops/argocd/image/register.sh
Comment (Roming22) [inline gitops/argocd/image/register.sh:101]: Put everything into functions so the script can be sourced without being executed.

File: gitops/credentials/kubeconfig/README.md
Comment (Roming22) [inline gitops/credentials/kubeconfig/README.md:3]: `aim at containing` -> `should contain`?

File: gitops/credentials/kubeconfig/README.md
Comment (Roming22) [inline gitops/credentials/kubeconfig/README.md:8]: Which is why I talk about encryption earlier. We should steer the user toward the right thing as much as possible.

**Review by Roming22 — CHANGES_REQUESTED** [review]

File: gitops/argocd/README.md
Comment (fgiloux) [reply gitops/argocd/README.md]: Looked at github and `kind`and `apiVersion` are not mentioned in the readme there either. Here is the [issue](https://github.com/kubernetes-sigs/kustomize/issues/738#issuecomment-459023753) where it is discussed to make them mandatory. My interpretation is that it got closed on the basis that these fields don't really bring value.

Comment (bnallapeta) [reply gitops/argocd/README.md]: I tried all the combinations locally. Here is what I found out.

kustomize build /path/to/folder/ 
works irrespective of whether we provide apiVersion and kind

kubectl apply -f /path/to/folder
with apiVersion and kind - no matches for kind kustomization
without apiVersion and kind - apiVersion and kind not set

kubectl apply -k /path/to/folder
works irrespective of whether we provide apiVersion and kind

So, in conclusion, does not seem to have any effect. My guess is with the 'kustomize build' and the '-k kubectl plugin', they have somehow ignored the kind and apiVersion.
I'm okay with both. If you think its neater and clearer for someone reading the file, we could keep the apiVersion and kind. Else, we could remove it.

File: gitops/argocd/image/register.sh
Comment (fgiloux) [reply gitops/argocd/image/register.sh]: I modified the script so that "context" is passed to the argocd command. The cluster name is still used as a key for the directory as the context may not be unique across multiple kubeconfigs

File: gitops/argocd/README.md
Comment (fgiloux) [reply gitops/argocd/README.md]: agreed. We could look at scaffolding later on if required

**Comment by fgiloux** [comment]
> I'm trying to split the configuration file per resource type. I used `plnsvc` to make it obvious that the kubeconfig referenced it the directory would host the Pipeline Service service. A more generic name would not necessarily convey that information.

I am not following what you are saying at all. You have used `plnsvc` to "contain" cluster level information. Here you are talking about service and today the directory contains kubeconfigs for accessing workload clusters. What are you calling a "service"? Providing "Pipelines service" is the whole purpose of this repository. It may help me with following your line of thought if you can give examples of siblings.

> * `/environment/plnsvc/cluster-name/config: do we need `config`? I would expect everything there to be configuration. What have you got in mind? Every cluster has 2 directories (`config`and`gitops`). For `plnsvc`clusters,`config`holds the information that ties the cluster to argocd, in the case of a`kcp`cluster there's additional information about tenants and namespaces. Those files holds the configuration that are the input to the pipeline_as_code process.`gitops` holds the argocd manifests.

Structures "should" be independent of the tooling used and reflects the nature of things as much as possible. In that sense there is currently in this PR "argocd-rbac" and "namespaces" as discussed with Bharath. Everything is already under a directory named "gitops".

File: gitops/argocd/README.md
Comment (fgiloux) [reply gitops/argocd/README.md:1]: I am fine with that. I will move The README under doc (and renamed it) and the files for the container image it under /images

File: gitops/credentials/kubeconfig/.gitignore
Comment (fgiloux) [reply gitops/credentials/kubeconfig/.gitignore:1]: I am not seeing the added value of an additional level and why this should be named template.

File: gitops/environment/plnsvc/base/argocd-rbac/argo-syncer-role.yaml
Comment (fgiloux) [reply gitops/environment/plnsvc/base/argocd-rbac/argo-syncer-role.yaml:5]: I am not following. I feel like it is reasonable to have a kcp component named syncer deployed under a namespace named kcp-syncer.

`kube-system` is definitely a bad choice.

Kubernetes documentation: "kube-system The namespace for objects created by the Kubernetes system"

Quoting David Eads: "kube-system contains service accounts which are used to run the kubernetes controllers. These service accounts are granted significant permissions (create pods anywhere, for instance). Since openshift builds on top of kube, we inherit the structure.
You should avoid putting anything "personal" in that namespace since kube considers it to be "owned" by kube and the permissions for the SAs inside are quite high"

`plnsvc` seems like a catch-all thing

File: gitops/argocd/README.md
Comment (fgiloux) [reply gitops/argocd/README.md]: I think that [this slide](https://speakerdeck.com/spesnova/introduction-to-kustomize?slide=73) is representative of a common practice with GitOps. Could you please share references for a different approach so that I can do some reading?

File: gitops/argocd/README.md
Comment (fgiloux) [reply gitops/argocd/README.md]: see comment above

File: gitops/argocd/README.md
Comment (fgiloux) [reply gitops/argocd/README.md]: I still would not put them in a public repository. We may look at integrating with [Vault](https://www.vaultproject.io/docs/platform/k8s) moving forward but that is out of scope of  this PR

File: gitops/argocd/image/register.sh
Comment (fgiloux) [reply gitops/argocd/image/register.sh]: I don't see how "plnsvc" is a good name for a cluster type in the context of this project

File: gitops/credentials/kubeconfig/README.md
Comment (fgiloux) [reply gitops/credentials/kubeconfig/README.md:8]: We need to do things step by step

File: gitops/credentials/kubeconfig/.gitignore
Comment (Roming22) [reply gitops/credentials/kubeconfig/.gitignore:1]: I would not expect this repository to be forked by an SRE team. They need only a subset of the information currently in this repository.
Therefore I would either:

- create a new repository with the filesystem structure that is expected for the gitops process;
- create a subfolder with the filesystem structure that is expected for the gitops process,

File: gitops/argocd/image/register.sh
Comment (Roming22) [reply gitops/argocd/image/register.sh]: I find `pipeline-service` a bit wordy. What do you propose?

**Comment by Roming22** [comment]
> Here you are talking about service and today the directory contains kubeconfigs for accessing workload clusters.

I'm not sure that's not entirely correct. My understanding is that technically we have 3 types/layers of clusters.

1. KCP: That's the proxy layer, accessed by our customers. Each environment has one and only one of those.
2. controllers: A "workloadcluster" that runs the controllers for a workspace (in the future multiple workspaces). The current implementation limits us to one controller per workloadcluster.
3. workers: A "workloadcluster" that can run pods. The "controllers cluster" may not be configured to run pods (although I would think that it does by default). The "workers" clusters are used to scale up the service when too many pods are being created and increasing the compute through a new cluster is necessary.

File: gitops/credentials/kubeconfig/.gitignore
Comment (fgiloux) [reply gitops/credentials/kubeconfig/.gitignore:1]: > I would not expect this repository to be forked by an SRE team.

That's a key point that we will need to discuss.

> Therefore I would either:
- create a new repository with the filesystem structure that is expected for the gitops process;
- create a subfolder with the filesystem structure that is expected for the gitops process,

 I am not closing any door. I feel however that it should not hold this PR.

File: gitops/environment/plnsvc/base/argocd-rbac/argo-syncer-role.yaml
Comment (fgiloux) [reply gitops/environment/plnsvc/base/argocd-rbac/argo-syncer-role.yaml:5]: yes that's correct. It is also what the argocd client generates when namespaces are specified with `argocd cluster add`

File: gitops/argocd/image/register.sh
Comment (fgiloux) [reply gitops/argocd/image/register.sh]: yes I did not mean pipeline-service. I had more in mind something, which describe the role of these clusters in the overall Pipelines-service. Something like:
- worker
- workloadcluster
- compute

File: gitops/argocd/image/register.sh
Comment (fgiloux) [reply gitops/argocd/image/register.sh]: The intent is to have the script run as part of a Tekton Pipeline, not really for it to get sourced but I can look at putting the code into functions tomorrow

**Comment by fgiloux** [comment]
> > Here you are talking about service and today the directory contains kubeconfigs for accessing workload clusters.
> 
> I'm not sure that's not entirely correct. My understanding is that technically we have 3 types/layers of clusters.
> 
>     1. KCP: That's the proxy layer, accessed by our customers. Each environment has one and only one of those.
> 
>     2. controllers: A "workloadcluster" that runs the controllers for a workspace (in the future multiple workspaces). The current implementation limits us to one controller per workloadcluster.
> 
>     3. workers: A "workloadcluster" that can run pods. The "controllers cluster" may not be configured to run pods (although I would think that it does by default). The "workers" clusters are used to scale up the service when too many pods are being created and increasing the compute through a new cluster is necessary.

I do agree on the 3 layers. Re 2 we will have very soon that kcp syncer is bound to a virtual workspace and as such serves many user workspaces. The phase 1 limitation is a single CRD version per workload cluster, which is acceptable for now. With phase 1 "controller clusters" are per design "worker clusters". With phase 2 this will be decoupled. We could run the controllers on separate clusters but as well decide to leverage kcp for them and run them in workspaces, hence on worker clusters but managed by kcp. As a consequence even if we do have the 3 theoretical types we may only use 2 of them.
To summarize we would have something like:
- gitops/environment/kcp/
- gitops/environment/compute/ (or worker, workload,...)
(possibly)
- gitops/environment/controllers (or operators)

**Comment by Roming22** [comment]
> To summarize we would have something like:
> 
> * gitops/environment/kcp/
> * gitops/environment/compute/ (or worker, workload,...)
>   (possibly)
> * gitops/environment/controllers (or operators)

Agreed.

**Comment by fgiloux** [comment]
@Roming22 @bnallapeta I think that I have applied the requested changes minus the "template" vs "fork", which we should discuss outside of this PR. Let me know whether it is good to merge or whether there are still things that need to be improved

**Review by Roming22 — APPROVED** [review]

**Comment by bnallapeta** [comment]
LGTM. 
@Roming22 @fgiloux : If we are ready to merge this, I will click that magic button. Please confirm :)

**Review by bnallapeta — APPROVED** [review]

**Comment by fgiloux** [comment]
Hi @bnallapeta Thanks. As Romain has already approved the PR you can go ahead and merge it

**Pull request merged**
